File: verify_srwma_computation.js
#!/usr/bin/env node
import { createRequire } from 'module';
import { readFileSync } from 'fs';
import { join } from 'path';
import { fileURLToPath } from 'url';
import { dirname } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const require = createRequire(import.meta.url);

// Load WASM module
const wasmPath = join(__dirname, 'pkg/my_project.js');
const wasm = require(wasmPath);

// Load test data (same as WASM benchmark)
const csvPath = join(__dirname, 'src/data/1MillionCandles.csv');
const content = readFileSync(csvPath, 'utf8');
const lines = content.trim().split('\n');
lines.shift(); // Skip header

// Parse close prices
const closes = [];
for (const line of lines) {
    const parts = line.split(',');
    if (parts.length >= 5) {
        closes.push(parseFloat(parts[4]));
    }
}

// Test with different sizes
const sizes = [1000, 10000, 100000];
const period = 14;

console.log('Verifying SRWMA computation...\n');

for (const size of sizes) {
    const data = new Float64Array(closes.slice(0, size));
    console.log(`\nData size: ${size}`);
    console.log(`First 5 values: ${Array.from(data.slice(0, 5))}`);
    console.log(`Last 5 values: ${Array.from(data.slice(-5))}`);
    
    // Run SRWMA
    const start = performance.now();
    const result = wasm.srwma_js(data, period);
    const end = performance.now();
    
    console.log(`\nComputation time: ${(end - start).toFixed(3)} ms`);
    console.log(`Result length: ${result.length}`);
    
    // Count NaN values (warmup period)
    let nanCount = 0;
    let firstNonNanIndex = -1;
    for (let i = 0; i < result.length; i++) {
        if (isNaN(result[i])) {
            nanCount++;
        } else if (firstNonNanIndex === -1) {
            firstNonNanIndex = i;
        }
    }
    
    console.log(`NaN count (warmup): ${nanCount}`);
    console.log(`First non-NaN index: ${firstNonNanIndex}`);
    console.log(`Expected warmup: ${period + 1}`);
    
    // Show some results
    if (firstNonNanIndex >= 0) {
        const firstValues = Array.from(result.slice(firstNonNanIndex, firstNonNanIndex + 5));
        console.log(`First 5 non-NaN values: ${firstValues.map(v => v.toFixed(6))}`);
        
        const lastValues = Array.from(result.slice(-5));
        console.log(`Last 5 values: ${lastValues.map(v => v.toFixed(6))}`);
    }
}

// Test fast API
console.log('\n\nTesting Fast API...');
const testData = new Float64Array(10000);
for (let i = 0; i < testData.length; i++) {
    testData[i] = closes[i] || 0;
}

// Allocate buffers
const len = testData.length;
const inPtr = wasm.srwma_alloc(len);
const outPtr = wasm.srwma_alloc(len);

// Copy data to WASM memory
const wasmMemory = new Float64Array(wasm.memory.buffer);
wasmMemory.set(testData, inPtr / 8);

// Run computation
const startFast = performance.now();
wasm.srwma_into(inPtr, outPtr, len, period);
const endFast = performance.now();

// Read results
const results = new Float64Array(wasm.memory.buffer, outPtr, len);
const resultCopy = Array.from(results);

console.log(`Fast API time: ${(endFast - startFast).toFixed(3)} ms`);
console.log(`First 5 values after warmup: ${resultCopy.slice(15, 20).map(v => v.toFixed(6))}`);

// Clean up
wasm.srwma_free(inPtr, len);
wasm.srwma_free(outPtr, len);

console.log('\nDone!');

File: src/bin/verify_srwma_computation.rs
use my_project::indicators::moving_averages::srwma::{srwma_with_kernel, SrwmaInput, SrwmaParams};
use my_project::utilities::data_loader::read_candles_from_csv;
use my_project::utilities::enums::Kernel;
use std::time::Instant;

fn main() {
	println!("Verifying SRWMA computation in Rust...\n");

	// Load the same CSV file
	let candles = read_candles_from_csv("src/data/1MillionCandles.csv").expect("Failed to load CSV");

	let sizes = [1000, 10000, 100000];
	let period = 14;

	for &size in &sizes {
		let data = &candles.close[..size.min(candles.close.len())];
		println!("\nData size: {}", size);
		println!("First 5 values: {:?}", &data[..5.min(data.len())]);
		println!("Last 5 values: {:?}", &data[data.len().saturating_sub(5)..]);

		let params = SrwmaParams { period: Some(period) };
		let input = SrwmaInput::from_slice(data, params);

		// Warmup
		for _ in 0..10 {
			let _ = srwma_with_kernel(&input, Kernel::Scalar);
		}

		// Time the computation
		let start = Instant::now();
		let result = srwma_with_kernel(&input, Kernel::Scalar).unwrap();
		let duration = start.elapsed();

		println!("\nComputation time: {:.3} ms", duration.as_secs_f64() * 1000.0);
		println!("Result length: {}", result.values.len());

		// Count NaN values
		let nan_count = result.values.iter().filter(|&&x| x.is_nan()).count();
		let first_non_nan = result.values.iter().position(|&x| !x.is_nan());

		println!("NaN count (warmup): {}", nan_count);
		println!("First non-NaN index: {:?}", first_non_nan);
		println!("Expected warmup: {}", period + 1);

		if let Some(idx) = first_non_nan {
			let first_values: Vec<String> = result.values[idx..idx + 5.min(result.values.len() - idx)]
				.iter()
				.map(|&v| format!("{:.6}", v))
				.collect();
			println!("First 5 non-NaN values: {:?}", first_values);

			let last_values: Vec<String> = result.values[result.values.len().saturating_sub(5)..]
				.iter()
				.map(|&v| format!("{:.6}", v))
				.collect();
			println!("Last 5 values: {:?}", last_values);
		}
	}

	// Test different kernels on 100k data
	println!("\n\nTesting different kernels on 100k data:");
	let data = &candles.close[..100000.min(candles.close.len())];
	let params = SrwmaParams { period: Some(period) };
	let input = SrwmaInput::from_slice(data, params);

	for kernel in [Kernel::Scalar, Kernel::Avx2, Kernel::Avx512, Kernel::Auto] {
		// Warmup
		for _ in 0..10 {
			let _ = srwma_with_kernel(&input, kernel);
		}

		let mut times = Vec::new();
		for _ in 0..100 {
			let start = Instant::now();
			let _ = srwma_with_kernel(&input, kernel).unwrap();
			times.push(start.elapsed().as_secs_f64() * 1000.0);
		}

		times.sort_by(|a, b| a.partial_cmp(b).unwrap());
		let median = times[times.len() / 2];
		let mean = times.iter().sum::<f64>() / times.len() as f64;

		println!("\nKernel {:?}:", kernel);
		println!("  Median: {:.3} ms", median);
		println!("  Mean: {:.3} ms", mean);
		println!("  Min: {:.3} ms", times[0]);
		println!("  Max: {:.3} ms", times[times.len() - 1]);
	}

	println!("\nDone!");
}
